File: tests/e2e/go-inventory/api_pass_test.go
// api_pass_test.go — Go counterpart to the Playwright console-pass spec.
//
// Uses unmodified cloud.google.com/go SDKs to provision the GCP resources
// directly against the simulator, then HTTP-pokes the running go-inventory
// app's endpoints and asserts correctness.
//
// Run with the app + simulator both reachable:
//
//	ENDPOINT=http://192.168.252.7:9000 APP_BASE=http://192.168.252.7:8081 \
//	  go test -v -run TestApiPass
//
// The test creates a Postgres DB via the Azure SQL path (same approach as the
// Playwright spec) since the simulator's Cloud SQL real-Postgres provisioning
// path returns connection strings the same way.
package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"
)

const (
	envEndpoint = "ENDPOINT"
	envAppBase  = "APP_BASE"

	tProject      = "inventory-app"
	tSecret       = "inventory-db-creds"
	tBucket       = "inventory-images"
	tTopic        = "inventory-events"
	tSub          = "inventory-worker"
	tKeyring      = "inventory-keyring"
	tKey          = "sku-key"
)

func endpoint() string { return envOr(envEndpoint, "http://192.168.252.7:9000") }
func appBase() string  { return envOr(envAppBase, "http://192.168.252.7:8081") }

func envOr(k, d string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return d
}

// ----------------------------------------------------------------------------
// TestApiPass: the whole flow as a single subtests-driven Go test.
// ----------------------------------------------------------------------------
func TestApiPass(t *testing.T) {
	ep := endpoint()
	t.Logf("simulator: %s", ep)
	t.Logf("app:       %s", appBase())

	var dbConn map[string]any

	t.Run("provision Postgres via Azure SQL path", func(t *testing.T) {
		if err := switchSpace(ep, "azure"); err != nil { t.Fatal(err) }

		api := "api-version=2023-08-01"
		base := ep + "/subscriptions/sub-e2e/resourceGroups/rg-e2e/providers/Microsoft.Sql/servers/inv-sql"

		mustHttp(t, "PUT", base+"?"+api,
			`{"location":"eastus","properties":{"administratorLogin":"azureadmin","administratorLoginPassword":"Password123!"}}`)
		mustHttp(t, "PUT", base+"/databases/inventory?"+api,
			`{"location":"eastus","properties":{}}`)

		body := mustHttp(t, "GET", base+"/databases/inventory?"+api, "")
		var resp struct {
			Properties struct {
				ConnectionInfo map[string]any `json:"connectionInfo"`
			} `json:"properties"`
		}
		if err := json.Unmarshal([]byte(body), &resp); err != nil { t.Fatal(err) }
		dbConn = resp.Properties.ConnectionInfo
		if dbConn == nil || !strings.Contains(fmt.Sprint(dbConn["engine"]), "PostgreSQL") {
			t.Fatalf("expected real Postgres connectionInfo, got: %s", body)
		}
		t.Logf("real Postgres DB ready: %v", dbConn["database"])
	})

	t.Run("provision GCP resources", func(t *testing.T) {
		if err := switchSpace(ep, "gcp"); err != nil { t.Fatal(err) }

		// 1. Secret Manager — create + addVersion.
		mustHttp(t, "POST", fmt.Sprintf("%s/v1/projects/%s/secrets?secretId=%s", ep, tProject, tSecret),
			`{"replication":{"automatic":{}}}`)

		secretBody := mustJSON(map[string]any{
			"url":      fmt.Sprintf("postgresql://%v:%v@%v:%v/%v",
				dbConn["user"], dbConn["password"], dbConn["host"], dbConn["port"], dbConn["database"]),
			"user":     dbConn["user"],
			"password": dbConn["password"],
		})
		dataB64 := base64.StdEncoding.EncodeToString([]byte(secretBody))
		mustHttp(t, "POST",
			fmt.Sprintf("%s/v1/projects/%s/secrets/%s:addVersion", ep, tProject, tSecret),
			fmt.Sprintf(`{"payload":{"data":"%s"}}`, dataB64))

		// 2. Storage bucket.
		mustHttp(t, "POST", fmt.Sprintf("%s/storage/v1/b?project=%s", ep, tProject),
			fmt.Sprintf(`{"name":"%s","location":"us-central1"}`, tBucket))

		// 3. Pub/Sub topic + subscription.
		mustHttp(t, "PUT", fmt.Sprintf("%s/v1/projects/%s/topics/%s", ep, tProject, tTopic), `{}`)
		mustHttp(t, "PUT", fmt.Sprintf("%s/v1/projects/%s/subscriptions/%s", ep, tProject, tSub),
			fmt.Sprintf(`{"topic":"projects/%s/topics/%s"}`, tProject, tTopic))

		// 4. KMS encrypt is implicit-on-use; nothing to provision.
		t.Logf("provisioned: secret + bucket + topic + subscription")
	})

	t.Run("app health=UP", func(t *testing.T) {
		deadline := time.Now().Add(90 * time.Second)
		var last string
		for time.Now().Before(deadline) {
			body, code, err := tryHttp("GET", appBase()+"/health", "")
			if err == nil && code == 200 {
				var h map[string]any
				if jerr := json.Unmarshal([]byte(body), &h); jerr == nil {
					last = body
					if fmt.Sprint(h["status"]) == "UP" {
						for _, svc := range []string{"db", "storage", "pubsub", "kms", "secret_manager"} {
							if v, _ := h[svc].(map[string]any); v == nil || v["ok"] != true {
								t.Fatalf("svc %s not OK in health: %s", svc, body)
							}
						}
						return
					}
				}
			}
			time.Sleep(2 * time.Second)
		}
		t.Fatalf("app /health never became UP; last=%s", last)
	})

	t.Run("POST /items round-trip", func(t *testing.T) {
		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
		defer cancel()
		_ = ctx
		body := mustHttp(t, "POST", appBase()+"/items",
			`{"name":"widget","sku":"W-001","stock":5}`)
		var item struct {
			ID int64 `json:"id"`
		}
		if err := json.Unmarshal([]byte(body), &item); err != nil { t.Fatal(err) }
		if item.ID <= 0 { t.Fatalf("expected id>0, got %d", item.ID) }

		listBody := mustHttp(t, "GET", appBase()+"/items", "")
		if !strings.Contains(listBody, "widget") {
			t.Fatalf("GET /items doesn't contain created item: %s", listBody)
		}

		// GET image → upload to bucket.
		imgBody := mustHttp(t, "GET", fmt.Sprintf("%s/items/%d/image", appBase(), item.ID), "")
		if !strings.Contains(imgBody, tBucket) {
			t.Fatalf("expected bucket %s in image URL: %s", tBucket, imgBody)
		}
	})
}

// ----------------------------------------------------------------------------
// helpers
// ----------------------------------------------------------------------------
func switchSpace(ep, provider string) error {
	body, _, err := tryHttp("GET", ep+"/api/spaces", "")
	if err != nil { return err }
	var resp struct {
		Spaces []struct {
			SpaceID  string `json:"space_id"`
			Provider string `json:"provider"`
		} `json:"spaces"`
	}
	if err := json.Unmarshal([]byte(body), &resp); err != nil { return err }
	for _, s := range resp.Spaces {
		if s.Provider == provider {
			_, _, err := tryHttp("POST", ep+"/api/spaces/"+s.SpaceID+"/switch", "")
			return err
		}
	}
	// create one
	created, _, err := tryHttp("POST", ep+"/api/spaces",
		fmt.Sprintf(`{"name":"e2e-%s","provider":"%s"}`, provider, provider))
	if err != nil { return err }
	var c struct{ SpaceID string `json:"space_id"` }
	if err := json.Unmarshal([]byte(created), &c); err != nil { return err }
	_, _, err = tryHttp("POST", ep+"/api/spaces/"+c.SpaceID+"/switch", "")
	return err
}

func mustHttp(t *testing.T, method, url, body string) string {
	t.Helper()
	out, code, err := tryHttp(method, url, body)
	if err != nil { t.Fatalf("%s %s: %v", method, url, err) }
	if code >= 500 { t.Fatalf("%s %s → HTTP %d: %s", method, url, code, out) }
	return out
}

func tryHttp(method, url, body string) (string, int, error) {
	var reader io.Reader
	if body != "" { reader = bytes.NewBufferString(body) }
	req, err := http.NewRequest(method, url, reader)
	if err != nil { return "", 0, err }
	if body != "" { req.Header.Set("Content-Type", "application/json") }
	resp, err := http.DefaultClient.Do(req)
	if err != nil { return "", 0, err }
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	return string(b), resp.StatusCode, nil
}

func mustJSON(v any) string {
	b, _ := json.Marshal(v)
	return string(b)
}

File: tests/e2e/java-orders/src/test/java/cloudlearn/orders/ApiPassTest.java
package cloudlearn.orders;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.eventbridge.EventBridgeClient;
import software.amazon.awssdk.services.kms.KmsClient;
import software.amazon.awssdk.services.kms.model.CreateKeyRequest;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.S3Configuration;
import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.CreateSecretRequest;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.CreateQueueRequest;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.time.Duration;

import static org.junit.jupiter.api.Assertions.*;

/**
 * API-pass: the JUnit5 counterpart to the Playwright console-pass spec.
 *
 *   - Uses unmodified aws-sdk-java-v2 (the same clients the app uses) to
 *     provision the 5 AWS resources directly against the simulator.
 *   - Then HTTP-pokes the running java-orders app's endpoints and asserts
 *     correctness.
 *
 * To run: have the app + simulator both reachable, then
 *
 *   ENDPOINT=http://192.168.252.7:9000 APP_BASE=http://192.168.252.7:8080 \
 *     mvn -Dtest=ApiPassTest test
 *
 * This test creates a Postgres DB via the Azure SQL path (since AWS RDS in
 * the simulator is metadata-only); the connection string + creds are stored
 * in Secrets Manager so the app reads them on startup.
 */
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class ApiPassTest {

    private static final String ENDPOINT = envOr("ENDPOINT", "http://192.168.252.7:9000");
    private static final String APP_BASE = envOr("APP_BASE", "http://192.168.252.7:8080");
    private static final String SECRET_NAME = "prod/orders/db";
    private static final String KMS_KEY     = "alias/orders-cc-key";
    private static final String BUCKET      = "orders-receipts";
    private static final String QUEUE_NAME  = "orders-processing-queue";

    private static final ObjectMapper M = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5)).build();

    private static StaticCredentialsProvider creds() {
        return StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test"));
    }

    @Test @Order(1)
    void provisionsRealPostgresViaAzureSqlPath() throws Exception {
        // Switch to an Azure space for the SQL create, then to AWS for the rest.
        switchToProviderSpace("azure");
        var api = "api-version=2023-08-01";
        var base = ENDPOINT + "/subscriptions/sub-e2e/resourceGroups/rg-e2e/providers/Microsoft.Sql/servers/orders-sql";

        // Server
        http("PUT", base + "?" + api, """
            {"location":"eastus","properties":{"administratorLogin":"azureadmin","administratorLoginPassword":"Password123!"}}""");
        // Database
        http("PUT", base + "/databases/orders?" + api, """
            {"location":"eastus","properties":{}}""");
        // GET-after-PUT surfaces connectionInfo
        var got = http("GET", base + "/databases/orders?" + api, null);
        JsonNode body = M.readTree(got.body());
        JsonNode conn = body.path("properties").path("connectionInfo");
        assertTrue(conn.path("engine").asText().contains("PostgreSQL"),
                "expected PostgreSQL engine in connectionInfo, got: " + body);

        // Switch back to AWS and store the JDBC URL in Secrets Manager.
        switchToProviderSpace("aws");
        String jdbcUrl = String.format("jdbc:postgresql://%s:%d/%s",
                conn.path("host").asText(),
                conn.path("port").asInt(),
                conn.path("database").asText());
        String secretJson = M.writeValueAsString(java.util.Map.of(
                "url", jdbcUrl,
                "user", conn.path("user").asText(),
                "password", conn.path("password").asText()));

        try (var sm = SecretsManagerClient.builder()
                .endpointOverride(URI.create(ENDPOINT))
                .region(Region.US_EAST_1)
                .credentialsProvider(creds()).build()) {
            try {
                sm.createSecret(CreateSecretRequest.builder()
                        .name(SECRET_NAME)
                        .secretString(secretJson)
                        .build());
            } catch (software.amazon.awssdk.services.secretsmanager.model.ResourceExistsException ignored) {
                sm.putSecretValue(b -> b.secretId(SECRET_NAME).secretString(secretJson));
            }
        }
    }

    @Test @Order(2)
    void provisionsKmsKey() {
        // AWS KMS CreateKey doesn't take a KeyId/alias — it generates one and
        // returns the metadata. Real AWS uses CreateAlias to attach an alias.
        // The simulator accepts whatever alias is passed at Encrypt time
        // (provisions Vault transit key lazily), so the description-only key
        // create here is enough — the app's first KMS.Encrypt with
        // alias/orders-cc-key triggers the lazy Vault key creation.
        try (var kms = KmsClient.builder()
                .endpointOverride(URI.create(ENDPOINT))
                .region(Region.US_EAST_1)
                .credentialsProvider(creds()).build()) {
            kms.createKey(CreateKeyRequest.builder()
                    .description("orders cc encryption")
                    .build());
        }
    }

    @Test @Order(3)
    void provisionsS3Bucket() {
        try (var s3 = S3Client.builder()
                .endpointOverride(URI.create(ENDPOINT))
                .region(Region.US_EAST_1)
                .credentialsProvider(creds())
                .serviceConfiguration(S3Configuration.builder()
                        .pathStyleAccessEnabled(true)
                        .chunkedEncodingEnabled(false).build())
                .build()) {
            try {
                s3.createBucket(CreateBucketRequest.builder().bucket(BUCKET).build());
            } catch (Exception ignored) { /* already exists */ }
        }
    }

    @Test @Order(4)
    void provisionsSqsQueue() {
        try (var sqs = SqsClient.builder()
                .endpointOverride(URI.create(ENDPOINT))
                .region(Region.US_EAST_1)
                .credentialsProvider(creds()).build()) {
            sqs.createQueue(CreateQueueRequest.builder().queueName(QUEUE_NAME).build());
        }
    }

    @Test @Order(5)
    void appHealthReturnsUp() throws Exception {
        // The app must already be running. Poll up to 90s for /health=UP.
        long deadline = System.currentTimeMillis() + 90_000;
        Exception last = null;
        while (System.currentTimeMillis() < deadline) {
            try {
                var r = http("GET", APP_BASE + "/health", null);
                if (r.statusCode() == 200) {
                    JsonNode h = M.readTree(r.body());
                    if ("UP".equals(h.path("status").asText())) {
                        for (String svc : new String[]{"db", "s3", "kms", "sqs", "eventbridge"}) {
                            assertTrue(h.path(svc).path("ok").asBoolean(),
                                    svc + " not OK: " + h.path(svc));
                        }
                        return;
                    }
                }
            } catch (Exception e) { last = e; }
            Thread.sleep(1500);
        }
        fail("app /health never became UP; last error: " + last);
    }

    @Test @Order(6)
    void postOrderRoundTrip() throws Exception {
        // POST /orders
        var post = http("POST", APP_BASE + "/orders",
                "{\"customer\":\"alice\",\"total_cents\":4999,\"cc\":\"4111111111111111\"}");
        assertEquals(200, post.statusCode(), "POST /orders body: " + post.body());
        JsonNode order = M.readTree(post.body());
        long orderId = order.path("id").asLong();
        assertTrue(orderId > 0);

        // GET /orders contains it
        var list = M.readTree(http("GET", APP_BASE + "/orders", null).body());
        boolean found = false;
        for (JsonNode o : list) if (o.path("id").asLong() == orderId) { found = true; break; }
        assertTrue(found, "newly-created order missing from GET /orders");

        // GET /orders/{id}/receipt → uploaded to S3
        var receipt = M.readTree(http("GET", APP_BASE + "/orders/" + orderId + "/receipt", null).body());
        assertTrue(receipt.path("receipt_url").asText().contains(BUCKET));
        assertTrue(receipt.path("size_bytes").asInt() > 0);

        // NATS inbox should have the OrderCreated event.
        var inbox = M.readTree(http("GET", ENDPOINT + "/__nats/inbox?prefix=aws.eventbridge.", null).body());
        boolean eventFound = false;
        for (JsonNode m : inbox.path("messages")) {
            if (m.toString().contains(String.valueOf(orderId))) { eventFound = true; break; }
        }
        assertTrue(eventFound, "OrderCreated event missing from NATS inbox");
    }

    // ---- helpers ----------------------------------------------------------
    private static String envOr(String k, String d) {
        String v = System.getenv(k); return (v == null || v.isEmpty()) ? d : v;
    }

    private static void switchToProviderSpace(String provider) throws Exception {
        var r = http("GET", ENDPOINT + "/api/spaces", null);
        JsonNode spaces = M.readTree(r.body()).path("spaces");
        String spaceId = "";
        for (JsonNode s : spaces) {
            if (provider.equals(s.path("provider").asText())) { spaceId = s.path("space_id").asText(); break; }
        }
        if (spaceId.isEmpty()) {
            var created = http("POST", ENDPOINT + "/api/spaces",
                    "{\"name\":\"e2e-" + provider + "\",\"provider\":\"" + provider + "\"}");
            spaceId = M.readTree(created.body()).path("space_id").asText();
        }
        http("POST", ENDPOINT + "/api/spaces/" + spaceId + "/switch", null);
    }

    private static HttpResponse<String> http(String method, String url, String body) throws Exception {
        var b = HttpRequest.newBuilder(URI.create(url)).timeout(Duration.ofSeconds(10));
        if (body != null) {
            b.method(method, HttpRequest.BodyPublishers.ofString(body));
            b.header("Content-Type", "application/json");
        } else {
            b.method(method, HttpRequest.BodyPublishers.noBody());
        }
        return HTTP.send(b.build(), HttpResponse.BodyHandlers.ofString());
    }
}
